### PR TITLE
Improve the performance of evidence handling and property retrieval

### PR DESCRIPTION
**Pull request by jwrosewell**

### CODE REVIEW

File: src/common-cxx
Comment (justadreamer) [inline src/common-cxx:1]: @jwrosewell could you please verify you pushed `common-cxx` changes to some remote branch, this commit reference does not seem to appear in the common-cxx, thus can't build this branch locally :( 
```sh
❯ git submodule update --init --recursive
fatal: remote error: upload-pack: not our ref eb436473c9c3422486f0c1c33a3c0e2a06752e16
Fetched in submodule path 'src/common-cxx', but it did not contain eb436473c9c3422486f0c1c33a3c0e2a06752e16. Direct fetching of that commit failed.
```

Comment (jwrosewell) [reply src/common-cxx:1]: It's [now ](https://github.com/51Degrees/common-cxx/tree/feature/performance-optimise)been pushed.

File: src/hash/hash.c [hunk 1]
```diff
@@ -1904,7 +1962,7 @@ size_t fiftyoneDegreesHashSizeManagerFromFile(
  	if (status == SUCCESS && EXCEPTION_OKAY) {
 		ResourceManagerFree(&manager);
 	}
-	else if (status != SUCCESS && EXCEPTION_OKAY) {
+	else if (exception != NULL && status != SUCCESS && EXCEPTION_OKAY) {
```
Comment (drasmart) [inline src/hash/hash.c:1965]: `exception` can not be both:
- not equal to NULL _(as added above)_
- equal to NULL _(as required by `EXCEPTION_OKAY` macro)_

```cpp
#define FIFTYONE_DEGREES_EXCEPTION_OKAY \
(exception == NULL || exception->status == FIFTYONE_DEGREES_STATUS_NOT_SET)
```
^ https://github.com/51Degrees/common-cxx/blob/71f3e1e4e19f9092d1cd341b7c296e0d8497e455/exceptions.h#L139-L143

Comment (jwrosewell) [reply src/hash/hash.c]: Code and comment modified. The intention of the method is purely to measure the amount of memory that is allocated during initialization. This is needed for environments that require modules to declare the amount of memory they're going to use during initialization. Therefore checking for the exception or throwing the exception are not needed. See commit https://github.com/51Degrees/device-detection-cxx/pull/87/commits/9cf57e5bd8567665ad4d07f62d9fd41cbb33b6c1

**Comment by justadreamer** [comment]
FYI Test failures reported by Max yesterday: 

image
image

**Comment by justadreamer** [comment]
has been incorporated into version/4.5

**Pull request closed without merging**